### PR TITLE
Api 1445 - Make array items collapsible

### PR DESCRIPTION
**Pull request by jocteau**
[API-1445](https://citizennet.atlassian.net/browse/API-1445)

The goal here is to make the items of an array collapsible so it is easier to move them around.
https://user-images.githubusercontent.com/61431714/120635670-0936d580-c46d-11eb-94b1-835407771896.mov

If needed we could have an expression for the items title instead of `Item {{ $index }}`.

One downside of doing this is that we attach `collapsed` to the item when saving the form.

@davezuch  If you have a better idea on how to do that, please let me know. I don't really know much about this library/repo.

Also,  I had no idea where to put the CSS so I added it here: 
https://github.com/citizennet/wildcat/pull/1241

### CODE REVIEW

File: src/directives/array.js [hunk 1]
```diff
@@ -123,6 +122,13 @@ angular.module('schemaForm').directive('sfArray', ['sfSelect', 'schemaForm', 'sf
             return formDefCache[index];
           };
 
+          scope.collapseForm = function() {
+            form.collapsed = !form.collapsed;
+            scope.modelArray.map(function(item) {
+              item.collapsed = form.collapsed
+            });
+          }
+
```
Comment (davezuch) [inline src/directives/array.js]:
```js
scope.collapsedItems = {};

scope.collapseForm = function() {
  form.collapsed = !form.collapsed;
  scope.modeArray.forEach(function(item, index) {
    scope.collapsedItems[index] = form.collapsed;
  });
};
```

File: src/directives/decorators/bootstrap/array.html [hunk 2]
```diff
@@ -3,16 +3,29 @@
...
   <ol class="list-group" ng-model="modelArray" ui-sortable="form.sortOptions">
     <li class="list-group-item {{form.fieldHtmlClass}}"
         ng-repeat="item in modelArray track by $index">
+      <div class="collapse-array-title" ng-hide="!form.collapsible">
+        <span ng-click="item.collapsed = !item.collapsed">
+            <i ng-hide="!item.collapsed" class="fa fa-caret-right"></i>
+            <i ng-hide="item.collapsed" class="fa fa-caret-down"></i>
+            Item {{$index + 1}}
+        </span>
+      </div>
```
Comment (davezuch) [inline src/directives/decorators/bootstrap/array.html]:
```js
<span ng-click="collapsedItems[$index] = !collapsedItems[$index]">
  <i ng-hide="!collapsedItems[$index]" class="fa fa-caret-right"></i>
  <i ng-hide="collapsedItems[$index]" class="fa fa-caret-down"></i>
```

File: src/directives/decorators/bootstrap/array.html [hunk 3]
```diff
@@ -3,16 +3,29 @@
...
       <button ng-hide="form.readonly || form.remove === null"
               ng-click="deleteFromArray($index)"
               type="button" class="close pull-right">
               <span aria-hidden="true">&times;</span><span class="sr-only">Close</span>
       </button>
-      <sf-decorator ng-init="arrayIndex = $index" form="copyWithIndex($index)"></sf-decorator>
+      <div uib-collapse="item.collapsed">
+        <sf-decorator ng-init="arrayIndex = $index" form="copyWithIndex($index)"></sf-decorator>
+      </div>
```
Comment (davezuch) [inline src/directives/decorators/bootstrap/array.html]:
```js
<div uib-collapse="collapsedItems[$index]">
```

**Review by davezuch — APPROVED** [review]
Looks good!

> @davezuch If you have a better idea on how to do that, please let me know. I don't really know much about this library/repo.

One solution that comes to mind is to have a dictionary on the scope with the collapsed state, and then toggle that. I'll provide inline comments for how I think that would work.

**Comment by jocteau** [comment]
> Looks good!
> 
> > @davezuch If you have a better idea on how to do that, please let me know. I don't really know much about this library/repo.
> 
> One solution that comes to mind is to have a dictionary on the scope with the collapsed state, and then toggle that. I'll provide inline comments for how I think that would work.

Thanks!
I thought about that too, but I didn't know how much I could add/change to the library since we forked it.

**Pull request merged**